compiler/aarch64/codegen/OMRMachine.cpp

```diff
@@ -79,18 +82,380 @@ TR::RealRegister *OMR::ARM64::Machine::freeBestRegister(TR::Instruction *current
                                                         TR::Register *virtualRegister,
                                                         TR::RealRegister *forced)
    {
-   TR_ASSERT(false, "Not implemented yet.");
+   TR::Register *candidates[NUM_ARM64_MAXR];
+   TR::Compilation *comp = self()->cg()->comp();
+   TR::MemoryReference *tmemref;
+   TR_BackingStore *location;
+   TR::RealRegister *best;
+   TR::Instruction *cursor;
+   TR::Node *currentNode = currentInstruction->getNode();
+   TR_RegisterKinds rk = (virtualRegister == NULL) ? TR_GPR : virtualRegister->getKind();
+   int numCandidates = 0;
+   int first, last;
+
+   if (forced != NULL)
+      {
+      best = forced;
+      candidates[0] = best->getAssignedRegister();
+      }
+   else
+      {
+      switch (rk)
+         {
+         case TR_GPR:
+            first = TR::RealRegister::FirstGPR;
+            last = TR::RealRegister::LastGPR;
+            break;
+         case TR_FPR:
+            first = TR::RealRegister::FirstFPR;
+            last = TR::RealRegister::LastFPR;
+            break;
+         default:
+            TR_ASSERT(false, "Unsupported RegisterKind.");
+            break;
+         }
+
+      for (int i = first; i <= last; i++)
+         {
+         TR::RealRegister *realReg = self()->getARM64RealRegister((TR::RealRegister::RegNum)i);
+         if (realReg->getState() == TR::RealRegister::Assigned)
+            {
+            candidates[numCandidates++] = realReg->getAssignedRegister();
+            }
+         }
+
+      cursor = currentInstruction;
+      while (numCandidates > 1 &&
+             cursor != NULL &&
+             cursor->getOpCodeValue() != TR::InstOpCode::label &&
+             cursor->getOpCodeValue() != TR::InstOpCode::proc)
+         {
+         for (int i = 0; i < numCandidates; i++)
+            {
+            if (cursor->refsRegister(candidates[i]))
+               {
+               candidates[i] = candidates[--numCandidates];
+               }
+            }
+         cursor = cursor->getPrev();
+         }
+      best = toRealRegister(candidates[0]->getAssignedRegister());
+      }
+
+   switch (rk)
+      {
+      case TR_GPR:
+         if (candidates[0]->getBackingStorage())
```

Comment:
Can the GPR and FPR cases be merged to save code space?  The logic is the same except for the potential spill sizes.

Reply:
Merged the GPR and FPR cases.